### PR TITLE
Look into spellchecking

### PR DESCRIPTION
**Pull request by plaindocs**
This is a quick look into spell checking the docs.

[Vale](https://errata-ai.gitbook.io/vale/) is a super extensible linter written in go. If you want to run it yourself:

1. [Install vale](https://errata-ai.gitbook.io/vale/getting-started/installation)
2. Run `vale pages`
3. Optionally install an editor plugin for [Atom](https://atom.io/packages/atomic-vale)

Work still to do:

- [x] clean up the word list in `vale/vocab.txt`
- [ ] ~make better exceptions (probably most of the `agent/v3/_help` dir, some html tags, some code that isn't detected as code, etc)~
- [x] Port changes to Agent https://github.com/buildkite/agent/pull/1224
- [x] add to CI :-p

CC @harrietgrace take a look, even if we're not yet good to go.

And if we ever need to figure out how to populate the vocab file again:

1. Delete `vale/vocab.txt`
2. Run vale without the custom work file :point_up: 
     * specify json output
     * grep out all matches with nasty regex
     * sort them and restrict to one of each match
     * strip out leading characters with nasty regex
     * strip out trailing characters with nasty regex
     * dump the lot into the vocab file

```
$ vale --output JSON pages/ | ack Match | sort | uniq | sed "s/\s*\"Match\": \"//" | sed "s/\"$//" > vale/vocab.txt
```

### CODE REVIEW

**Comment by plaindocs** [comment]
Everything except ba2ec91 and 4842023 are actual typo fixes. I'll cherry pick them out if this looks like it is going to hang around long.

:eyes: especially welcome on `vale/vocab.txt` in case I've missed anything obvious.

File: pages/apis/rest_api.md.erb [hunk 1]
```diff
@@ -105,7 +105,7 @@ The following changes were made in v2 of our API:
 
 To make getting started easier, check out these clients available from our contributors:
 
- * [Buildkit](https://github.com/Shopify/buildkit) for [Ruby](https://www.ruby-lang.org)
+ * [Buildkite](https://github.com/Shopify/buildkit) for [Ruby](https://www.ruby-lang.org)
```
Comment (plaindocs) [inline pages/apis/rest_api.md.erb:108]:
```suggestion
 * [Buildkit](https://github.com/Shopify/buildkit) for [Ruby](https://www.ruby-lang.org)
```

File: pages/pipelines/conditionals.md.erb [hunk 2]
```diff
@@ -59,7 +59,7 @@ The below expressions are supported by the `if` attribute:
  			<td><code>'feature-branch' "feature-branch"</code></td>
  		</tr>
 		<tr>
- 			<td>Booleans</td>
+			<td>Boolean</td>
```
Comment (matthewd) [inline pages/pipelines/conditionals.md.erb:62]: "Boolean literals"? Or even "Literals" and fold in `null` as well? Seems a bit odd that that one's singular too.

File: pages/agent/v3/help/_bootstrap.txt [hunk 3]
```diff
@@ -52,7 +52,7 @@ Options:
    --build-path value                   Directory where builds will be created [$BUILDKITE_BUILD_PATH]
    --hooks-path value                   Directory where the hook scripts are found [$BUILDKITE_HOOKS_PATH]
    --plugins-path value                 Directory where the plugins are saved to [$BUILDKITE_PLUGINS_PATH]
-   --command-eval                       Allow running of arbitary commands [$BUILDKITE_COMMAND_EVAL]
+   --command-eval                       Allow running of arbitrary commands [$BUILDKITE_COMMAND_EVAL]
```
Comment (matthewd) [inline pages/agent/v3/help/_bootstrap.txt:55]: Presumably this (and the next) suggests a change in the agent too

File: vale/vocab.txt [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+Explainshell
+fff
+ffxdq
+Firewalled
+Flexbox
+flickr
+gcloud
+gif
+github
+Github
```
Comment (matthewd) [inline vale/vocab.txt:49]: GitHub

File: vale/vocab.txt [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+Basecamp
+Basscss
+Bazel
+Bitium
+boolean
+boomper
+Boomper
+Buildbox
+buildkite
+Buildkite
```
Comment (matthewd) [inline vale/vocab.txt:20]: Per the other change, we probably need `buildkit` in here.

File: vale/vocab.txt [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+buildscript
+chatbot
+chroot
+chroots
+cli
+cloudformation
+Cloudwatch
```
Comment (matthewd) [inline vale/vocab.txt:27]: These should probably be CloudWatch and CloudFormation (though the all-lower variant might be what we want somewhere?)

File: vale/vocab.txt [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+fff
+ffxdq
+Firewalled
+Flexbox
+flickr
+gcloud
+gif
+github
+Github
+Gitlab
```
Comment (matthewd) [inline vale/vocab.txt:50]: GitLab

File: vale/vocab.txt [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+globbing
+golang
+Gradle
+gz
+Hashicorp
+heroku
+hostname
+http
+https
```
Comment (matthewd) [inline vale/vocab.txt:59]: Are a lot of these entries from URLs? Or just code blocks? I see a lot of very lowercase words I wouldn't expect to turn up in prose.

**Comment by harrietgrace** [comment]
So psyched for this! It's looking great 🌈

Omg so many spelling mistakes 😱

File: vale/vocab.txt (hunk 5 above)
Comment (plaindocs) [reply vale/vocab.txt:20]: Heh yeah, I thought that was a typo and removed it, and snuck it back in in e7b7289

**Comment by plaindocs** [comment]
Hey @matthewd thanks for the :eyes: 

What is happening is that the vale spell check comes with built in "filters for acronyms, abbreviations, and numbers." and *think* that includes camel case stuff.

So probably some code in there is getting treated as text. I suspect the help files. For example, removing `http` from the vocab list gives us

```
 pages/agent/v3/help/_artifact_upload.txt
 32:39  error  Did you really mean 'http'?  Vale.Spelling 


 pages/integrations/cc_menu.md.erb
 47:1  error  Did you really mean 'http'?  Vale.Spelling 
 51:1  error  Did you really mean 'http'?  Vale.Spelling 

✖ 3 errors, 0 warnings and 0 suggestions in 152 files.
```

The CC_menu page needs refactoring, and it looks like the indented code block in the help pages indeed are showing up as text.

So over all, it is a slight chicken egg situation with needing to get the amount of errors down to something manageable to be able to look at them.

I'll try to remove a ton more of the stuff that looks like URL contents and see what gets flagged. Overall, keeping the vocal list as small as possible is best I think.

And I need to make sure the case sensitive stuff works as well.

**Comment by plaindocs** [comment]
Yeah, looks like the blocks in the help aren't indented right to be detected as code. Making it four spaces instead of one tab and one space makes them not turn up.

**Comment by plaindocs** [comment]
There is probably a better way than pulling the installer every time.

File: vale/vocab.txt [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+Memcache
+Minitest
+mutex
+namespace
+negator
+Noreply
+npm
+nssm
+Okta
+Pagerduty
```
Comment (JuanitoFatas) [inline vale/vocab.txt:73]: We can add OAuth here :)

Comment (plaindocs) [reply vale/vocab.txt:73]: `oAuth` doesn't show up as a misspelling (probably because camel case).

I'll add it a different check.

Comment (plaindocs) [reply vale/vocab.txt:73]: Thanks @JuanitoFatas !

**Comment by plaindocs** [comment]
Alright, after a long rambling PR mostly talking to myself, I'm ready to merge this in.
I'll track proposed enhancements to the linter separately.

**Pull request merged**